### PR TITLE
[go_router] Avoid logging when `debugLogDiagnostics` is `false`

### PR DESCRIPTION
**Pull request by ValentinVignal**
Fixes https://github.com/flutter/flutter/issues/134187

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/logging.dart [hunk 1]
```diff
@@ -4,22 +4,35 @@
...
 /// The logger for this package.
-final Logger log = Logger('GoRouter');
+@visibleForTesting
+final Logger logger = Logger('GoRouter');
+
+/// Whether or not the logging is enabled.
+bool _enabled = false;
+
+/// Logs the message if logging is enabled.
+void log(String message) {
```
Comment (chunhtai) [inline packages/go_router/lib/src/logging.dart:19]: use `log` may be ambiguous if we later add other level for some reason. 

should this just set logger.level = Level.OFF when set enabled?

Comment (ValentinVignal) [reply packages/go_router/lib/src/logging.dart:19]: > use log may be ambiguous if we later add other level for some reason.

We could later on change the signature to

```dart
void log(String message, [Level level = Level.INFO]);
```

> should this just set logger.level = Level.OFF when set enabled = false?

This works too.
Unfortunately, doing so would force the user to set `hierarchicalLoggingEnabled` to `true`. If not, `logging` throws:

```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following UnsupportedError was thrown building MyApp(dirty):
Unsupported operation: Please set "hierarchicalLoggingEnabled" to true if you want to change the
level on a non-root logger.
```

I'm not sure that acceptable to for the user to do that?

Comment (chunhtai) [reply packages/go_router/lib/src/logging.dart:19]: I see, yeah this is fine

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**